cluster/etcd_service_discovery.go

```diff
@@ -267,10 +274,15 @@ func (sd *etcdServiceDiscovery) getServerFromEtcd(serverType, serverID string) (
 
 // GetServersByType returns a slice with all the servers of a certain type
 func (sd *etcdServiceDiscovery) GetServersByType(serverType string) (map[string]*Server, error) {
-	if m, ok := sd.serverMapByType.Load(serverType); ok {
-		sm := m.(map[string]*Server)
-		if len(sm) > 0 {
-			return sm, nil
+	sd.mapByTypeLock.RLock()
+	defer sd.mapByTypeLock.RUnlock()
+	if m, ok := sd.serverMapByType[serverType]; ok {
+		if len(m) > 0 {
+			ret := make(map[string]*Server, len(sd.serverMapByType))
+			for k, v := range sd.serverMapByType[serverType] {
```

Comment:
It is not obvious why your are iterating over a map to build the exact same map. I have the context by what we've talked and I had to think for a while before remembering it. I'd leave a comment here explaining why this is necessary.

Reply:
Good idea